src/training/graph_group.h

```diff
@@ -58,19 +65,32 @@ class AsyncGraphGroup : public GraphGroup {
         return;
 
       // @TODO read guard on parameters
-      std::lock_guard<std::mutex> guard(sync_);
-      oldParams->copyFrom(params_);
+      int pos = 0;
+      for (int idx = 0; idx < devices_.size(); idx++) {
+        std::lock_guard<std::mutex> guard( shardSync_[idx] );
+        oldParams->subtensor(pos , params_[idx]->size())->copyFrom(params_[idx]);
+        pos += shardSize_;
```

Comment:
I was using --amend. Shouldn't have done that.

The working one was using cudaStreamSynchronize(0) in pushGradient, and using std::async instead of thread. 

Have you tried my latest pull with thread and just fix the mutex construction?